doc/source/dev/toolchain.rst

```diff
@@ -0,0 +1,136 @@
+
+Python Versions
+---------------
+
+SciPy is compatible with several versions of Python.
+[*NOTE: The level of support and dates is currently under discussion.
+This table shows what it could look like.*]
+
+================  =================================================
+ Date             Pythons supported
+================  =================================================
+ <= 2018           Py2.7, Py3.5 - Py3.7
+ 2019              Py3.5 - (but Py2.7-specific code not removed)
+ 2020              Py3.?- (removal of Py2.7-specific code permitted)
+================  =================================================
+
+
+C Compilers
+-----------
+SciPy is compatible with most modern C compilers.  However CPython on Windows is
+built with specific versions of the Microsoft Visual C++ compiler [1]_, as is the
+corresponding build of SciPy.  This has implications for the C language standards
+that can be supported.
+
+===================   ==============   ===================
+CPython               MS Visual C++    C Standard
+===================   ==============   ===================
+3.5, 3.6                14.0           C90 + most of C99
+3.3, 3.4                10.0           C90 + some of C99
+2.7, 3.0, 3.1, 3.2       9.0           C90
+===================   ==============   ===================
+
+
+C Language Standards
+--------------------
+[*NOTE: I'm unaware of any official SciPy decisions on the support of C standards.
+This table shows what it could look like.*]
+
+================  =========================================
+ Date              C Standard
+================  =========================================
+ <= 2018           C90
+ 2019              C90 for old code, may allow C99 for new
+ 2020              C99
+ No decision       C11
+ No decision       C17, C18
+================  =========================================
+
+Some C99 features would be useful for scientific programming, in particular better support of
+IEEE 754 [2]_.  Experience has shown that not all features are supported equally well across
+all platforms, SciPy has a small include file ``scipy/_lib/_c99compat.h`` which provides
+access to a few functions.  Use in conjunction with ``<numpy/npy_math.h>``.
+
+================================  ========================================================
+ Feature                           Workaround
+================================  ========================================================
+ isnan(), isinf(), isfinite()      Use sc_isnan(), sc_isinf(), sc_isfinite()
+ NAN                               Use NPY_NAN (*almost* equivalent)
+ inline functions                  Make static functions and place in an include .h file
+ mid-block variable declarations   Declare variables at the top of block
+================================  ========================================================
+
+Fortran Compilers
+-----------------
+
+[*What is the requirement on a Fortran compiler*]
```

Comment:
Any well-maintained compiler I think, which is basically: gfortran (>=4.8.0 I think) , ifort, flang.